Sources/NIOPosix/Bootstrap.swift

```diff
@@ -521,6 +521,200 @@ public final class ServerBootstrap {
+    ///                    )
+    ///                )
+    ///            }
+    ///        }
+    ///    ) { channel in
+    ///        print("Handling new connection")
+    ///        await self.handleConnection(channel: channel)
+    ///        print("Done handling connection")
+    ///    }
+    ///
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - target: The ``BindTarget`` to use.
+    ///   - serverBackPressureStrategy: The back pressure strategy used by the server socket channel.
+    ///   - childChannelInitializer: A closure to initialize the channel. The return value of this closure is used in the `onConnection`
+    ///                              closure.
+    ///   - onConnection: A closure to handle the connection. Use the channel's `inbound` property to read from
+    ///                   the connection and channel's `outbound` to write to the connection.
+    ///
+    /// - Note: If the server is not closed using a closure mechanism like above, the bind method must be cancelled using task
+    ///         cancellation for the method to return. Task cancellation will force close the server and wait for all remaining
+    ///         sub task (connection callbacks) to finish.
+    @available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+    public func bind<Inbound: Sendable, Outbound: Sendable>(
+        target: BindTarget,
+        serverBackPressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark? = nil,
+        childChannelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<NIOAsyncChannel<Inbound, Outbound>>,
+        _ onConnection: @escaping @Sendable (
+            _ channel: NIOAsyncChannel<Inbound, Outbound>
+        ) async throws -> ()
+    ) async throws {
+        let channel = try await self.makeConnectedChannel(
+            target: target,
+            serverBackPressureStrategy: serverBackPressureStrategy,
+            childChannelInitializer: childChannelInitializer
+        )
+
+        try await withTaskCancellationHandler {
+            try await channel.executeThenClose { inbound, outbound in
+                // we need to dance the result dance here, since we can't throw from the
+                // withDiscardingTaskGroup closure.
+                let result = await withDiscardingTaskGroup { group -> Result<Void, any Error> in
+                    do {
+                        try await channel.executeThenClose { inbound in
+                            for try await connectionChannel in inbound {
+                                group.addTask {
+                                    do {
+                                        try await connectionChannel.executeThenClose { _, _ in
+                                            try await onConnection(connectionChannel)
+                                        }
+                                    } catch {
+                                        // ignore single connection failures?
```

Comment:
Yeah this is a bit of a though one. I'm not sure what the right thing to do is here. A single connection failure shouldn't bring down the whole server. I think a good first step might be to make the `onConnection` closure non-throwing so that users are forced to handle errors themselves.